### ISSUE TITLE
opening jupytext notebook with a base_url is failing

### ISSUE THREAD

**Issue by parmentelat** [jupyter/notebook#6942]
## Description

possibly related to #6914

getting a 404 error when using the *Open with* → *Notebook* menu on a jupytext notebook

![nb7-text-base-url](https://github.com/jupyter/notebook/assets/4519227/e34b8fb4-5fd1-468d-8d06-8a6cab9f990e)

this capture was made when running

```bash
jupyter lab --NotebookApp.base_url=/jlab4-nb7-text/
```

the 2 first first attempts work fine:
* on a ipynb, *Open with* → *Notebook*
* on a jupytext, *Open with* → *Jupytext Notebook*

the video may not be super clear but in both these cases the new tab's URL contains the base_url 

**BUT**
* on a jupytext, *Open with* → *Notebook* ends up on a 404 because the base_url thingy was not inserted in the URL

## Reproduce

see above; this is with
```
pip freeze | egrep 'jupyte|notebook'
jupyter-events==0.6.3
jupyter-lsp==2.2.0
jupyter_client==8.2.0
jupyter_core==5.3.1
jupyter_server==2.6.0
jupyter_server_terminals==0.4.4
jupyterlab==4.0.2
jupyterlab-pygments==0.2.2
jupyterlab-widgets==3.0.7
jupyterlab_jupytext @ git+https://github.com/mwouts/jupyterlab-jupytext.git@d451f5c6fddc958bd3b7082720a7627aa4681fbd
jupyterlab_server==2.23.0
jupytext==1.14.6
notebook==7.0.0rc1
notebook_shim==0.2.3
```

## Context

<details><summary>Command Line Output</summary>
<pre>
[W 2023-06-22 12:55:39.037 ServerApp] 404 GET /edit/tests/sample-jupytext.md?factory=Notebook (cddb81bae87041cc81171cb9a9812c73@::1) 1.64ms referer=http://localhost:8891/jlab4-nb7-text/tree/tests
</pre>
</details>

<details><summary>Browser Output</summary>

![image](https://github.com/jupyter/notebook/assets/4519227/b1630b9c-35af-475a-9ab0-a69716a13648)

</details>

**Comment by jtpio** [comment]
Thanks @parmentelat for reporting.

Another odd thing when looking at the screencast above is that when selecting `Open With > Notebook`, then it should redirect to `<base-url>/notebooks` and not `<base-url>/edit`.

Maybe it's an issue with this handler: 

https://github.com/jupyter/notebook/blob/8efd94a0a8234d0079752e7af0076dc1b271339e/notebook/app.py#L332-L337

**Comment by parmentelat** [comment]
with the change from https://github.com/jupyter/notebook/actions/runs/5345232824?pr=6943, what I see for the URLs is this

```
# Open with → Notebook
http://localhost:8889/jlab4-nb7rc2-text/edit/tests/sample-jupytext.md?factory=Notebook
# Open with → Jupytext Notebook, I get simply this
http://localhost:8889/jlab4-nb7rc2-text/edit/tests/sample-jupytext.md
```

as you can see in both cases the URL contains `/edit/`, while the plain ipynb notebook contains `/notebooks/` like so
```
http://localhost:8889/jlab4-nb7rc2-text/notebooks/tests/sample-notebook.ipynb
```

**Comment by jtpio** [comment]
This is likely because of the folloiwing regex if the file extension is `.md`.

https://github.com/jupyter/notebook/blob/8efd94a0a8234d0079752e7af0076dc1b271339e/notebook/app.py#L334

**Comment by parmentelat** [comment]
I am also seeing another couple differences between the notebook views for a ipynb and a jupytext - focusing on md:myst for now:

- the ipynb view has 2 additional menu entries `Run` and `Kernel`
- and, much more cosmetic, there's an additional vertical space on top of the jupytext view

see the screenshot below

I guess it might be related to the `/notebooks/` *vs* `/edit/` diff in the URL ?

![image](https://github.com/jupyter/notebook/assets/4519227/2d2c6c66-947a-42bb-b5c4-85af5ca39b5b)

**Comment by jtpio** [comment]
> I guess it might be related to the `/notebooks/` _vs_ `/edit/` diff in the URL ?

Yes. The `/notebooks` page has some CSS overrides so the notebook looks like the classic notebook. And the `/edit` page disables some menu items:

https://github.com/jupyter/notebook/blob/8efd94a0a8234d0079752e7af0076dc1b271339e/packages/application-extension/src/index.ts#L269-L272

Maybe the proper fix would be to make sure the Jupytext notebooks are open with `/notebooks`, at least when choosing the `Notebook` factory.

Not sure if this redirect to `/edit` would still be needed then. I think it was added in RetroLab a while ago, would need to find the related PR for more context.

**Comment by parmentelat** [comment]
> Maybe the proper fix would be to make sure the Jupytext notebooks are open with /notebooks, at least when choosing the Notebook factory.

totally agreed :)

**Comment by jtpio** [comment]
> Not sure if this redirect to `/edit` would still be needed then. I think it was added in RetroLab a while ago, would need to find the related PR for more context.

It was added in https://github.com/jupyterlab/retrolab/pull/98

**Comment by parmentelat** [comment]
I gave that a quick try within the same conda env
(using `jupyter notebook file` instead because of course `Jupyter command `jupyter-classic` not found.`)

```
jupyter notebook sample-notebook.ipynb
```

yields this

<img width="1585" alt="image" src="https://github.com/jupyter/notebook/assets/4519227/db0b55e5-324b-4f47-80fc-fd4663e9c9e9">

and same with the other sample notebooks, so that feature seems to have grown broken in the meanwhile, has it not ?

**Comment by jtpio** [comment]
> and same with the other sample notebooks, so that feature seems to have grown broken in the meanwhile, has it not ?

Just tried locally and it seems to be working:

- With a notebook it opens under `/notebooks`
- With another file it opens under `/edit`

**Comment by jtpio** [comment]
One idea could be to modify the regex so the redirect handler does not apply when the URL contains `?factory=Notebook`:

https://github.com/jupyter/notebook/blob/8efd94a0a8234d0079752e7af0076dc1b271339e/notebook/app.py#L334

**Comment by parmentelat** [comment]
so, 7.0.0rc2 does address the initial concern, in that I no longer get a 404 and the notebook does open

however as you noticed, the `/edit/` path is new as compared to the classic notebook experience
I believe @jtpio had tried to get rid of this discrepancy at some point
imho it would be great if we could achieve this, has the presence of 2 different paths has potential harmful side effects:
- it becomes possible to have 2 tabs opened on the same notebook, but that don't know they share the same file; I'm not exactly sure about that, I am no expert jupyterlab user (yet), but it feels wrong
- also external toolings may be confused if they rely on the fact that the current URL contains `/notebooks/` in order to parse for the notebook location or similar

so to summarize, I can live with what 7.0.0rc2 gives us, but I quite agree that there remains possible improvements
